internal/persistence/postgres/pool.go

```diff
@@ -0,0 +1,154 @@
+package postgres
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"go.uber.org/zap"
+)
+
+var pools = map[string]PoolOpenFunc{}
+
+// Pool is the interface of the pgxpool.Pool type, used by the postgres
+// persistence layer to query the database. A custom pool type can be
+// registered using `RegisterPool` if it implements this interface.
+type Pool interface {
+	Close()
+	Acquire(ctx context.Context) (*pgxpool.Conn, error)
+	AcquireFunc(ctx context.Context, f func(*pgxpool.Conn) error) error
+	AcquireAllIdle(ctx context.Context) []*pgxpool.Conn
+	Config() *pgxpool.Config
+	Stat() *pgxpool.Stat
+	Exec(ctx context.Context, sql string, arguments ...interface{}) (pgconn.CommandTag, error)
+	Query(ctx context.Context, sql string, args ...interface{}) (pgx.Rows, error)
+	QueryRow(ctx context.Context, sql string, args ...interface{}) pgx.Row
+	QueryFunc(ctx context.Context, sql string, args []interface{}, scans []interface{},
+		f func(pgx.QueryFuncRow) error) (pgconn.CommandTag, error)
+	SendBatch(ctx context.Context, b *pgx.Batch) pgx.BatchResults
+	Begin(ctx context.Context) (pgx.Tx, error)
+	BeginTx(ctx context.Context, txOptions pgx.TxOptions) (pgx.Tx, error)
+	BeginFunc(ctx context.Context, f func(pgx.Tx) error) error
+	BeginTxFunc(ctx context.Context, txOptions pgx.TxOptions, f func(pgx.Tx) error) error
+	CopyFrom(ctx context.Context, tableName pgx.Identifier, columnNames []string, rowSrc pgx.CopyFromSource) (int64, error)
+	Ping(ctx context.Context) error
+}
+
+// PoolOpenFunc defines a function that can instantiate a Pool instance with the given options.
+type PoolOpenFunc func(opts Opts, logger *zap.Logger) (Pool, error)
+
+type PoolOpts struct {
+	// Optional The name of the postgres pool implementation to use.
+	// By default, only `pgx` is available.
+	Name string
+
+	// MaxConns is the maximum size of the pool. Default to `persistence.DefaultMaxConn`.
+	MaxConns int32
+
+	// MinConns is the minimum size of the pool. Default to `persistence.DefaultMinConn`.
+	MinConns int32
+
+	// MaxConnLifetime is the duration since creation after which a connection will
+	// be automatically closed. Default to `persistence.DefaultMaxConnLifetime`.
+	MaxConnLifetime time.Duration
+
+	// MaxConnIdleTime is the duration after which an idle connection will be automatically
+	// closed by the health check. Default to `persistence.DefaultMaxConnLifetime`.
+	MaxConnIdleTime time.Duration
+
+	// HealthCheckPeriod is the duration between checks of the health of
+	// idle connections. Default to `persistence.DefaultHealthCheckPeriod`.
+	HealthCheckPeriod time.Duration
+
+	// ReadOnly indicates the pool will be used for read only operations.
+	ReadOnly bool
+}
```

Comment:
Building a new PoolOpts  using `PoolOpts{}` will give us an invalid default object, due to the zero-value of these types (see PoolOpts.Validate()). This is not ideal. Should I create a proper `newPooOpts()` to avoid this? (not sure yet how this would work since we also use `PostgresOpts{}` directly.

Reply:
yeah, I wrote this comment prior the addition of the pool validation function: https://github.com/Kong/koko/pull/502/files#diff-d18f5e36820a1a2d869d4c377f9d3bc3ee07bd24391f4c43ec7e7a6b80e1b5f9R71

so with the default yaml settings being applied at config loading.. and also koko calling this Validate function, the normal path is fine.

Reply:
will fix!